walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,90 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
 
 public final class FindMeetingQuery {
+
+  /** First Solution: 
+   * Sort {@code events} by each event's starting time
+   * Filter out any events that do not include attendees in the request
+   * Map the filter eventlist to a list of each event's {@code TimeRange}
+   * Find vacant time slots with a duration longer than or equal to the request
+   * duration and add it to the {@code Collection<TimeRange>}
+   *
+   * This solution is O(nlog_n) becuase of the use of the stream.sort() method. 
+   * Although, all of the test cases use Event Collections that are 
+   * already in order, which would allow me to remove .sort() and 
+   * lower the algorithmic complexity to O(n).
+   * But it's not safe to assume the Event Collection will already be in order 
+   * before the method call.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // If the request has no attendees, the open time slot is the entire day.
+    if (request.getAttendees().isEmpty()) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    // Sort {@code events} by starting time.
+    List<Event> eventsList = new ArrayList(events);
+    eventsList.sort((Event a, Event b) -> {
+        return TimeRange.ORDER_BY_START.compare(a.getWhen(), b.getWhen());
+    });
+
+    // Filter out attendees not in {@code request}.
+    eventsList = eventsList.stream()
+                           .filter(e -> request.getAttendees()
+                                               .containsAll(e.getAttendees()))
```

Comment:
I think you want to consider any events that contain any of the attendees instead of only events that contain all of the attendees.

Reply:
Oh you're right! There weren't any tests with events that had additional attendees so I hadn't realized.